File: java-app/src/main/java/com/showcase/java/domain/command/CommandHandler.java
package com.showcase.java.domain.command;

import com.showcase.java.domain.bus.EventBus;
import com.showcase.java.domain.event.DomainEvent;
import com.showcase.java.domain.store.EventStore;

import java.time.Instant;
import java.util.List;
import java.util.Set;

public class CommandHandler {

    private final EventStore eventStore;
    private final EventBus eventBus;

    public CommandHandler(EventStore eventStore, EventBus eventBus) {
        this.eventStore = eventStore;
        this.eventBus = eventBus;
    }

    public List<DomainEvent> handle(Command command) {
        var events = switch (command) {
            case Command.CreateTask c -> handleCreate(c);
            case Command.AssignTask c -> handleAssign(c);
            case Command.ChangeTaskPriority c -> handleChangePriority(c);
            case Command.CompleteTask c -> handleComplete(c);
            case Command.AddTagToTask c -> handleAddTag(c);
            case Command.MoveTaskToProject c -> handleMoveToProject(c);
        };
        eventStore.append(command.aggregateId(), events);
        eventBus.publish(events);
        return events;
    }

    private List<DomainEvent> handleCreate(Command.CreateTask cmd) {
        var now = Instant.now();
        return List.of(new DomainEvent.TaskCreated(
                cmd.taskId().value(),
                now,
                cmd.title(),
                cmd.description(),
                cmd.priority(),
                cmd.dueDate(),
                cmd.tags() != null ? cmd.tags() : Set.of()
        ));
    }

    private List<DomainEvent> handleAssign(Command.AssignTask cmd) {
        return List.of(new DomainEvent.TaskAssigned(
                cmd.taskId().value(),
                Instant.now(),
                cmd.assigneeId()
        ));
    }

    private List<DomainEvent> handleChangePriority(Command.ChangeTaskPriority cmd) {
        return List.of(new DomainEvent.TaskPriorityChanged(
                cmd.taskId().value(),
                Instant.now(),
                cmd.newPriority(), // simplified — real impl would load current priority
                cmd.newPriority()
        ));
    }

    private List<DomainEvent> handleComplete(Command.CompleteTask cmd) {
        var now = Instant.now();
        return List.of(new DomainEvent.TaskCompleted(
                cmd.taskId().value(),
                now,
                now
        ));
    }

    private List<DomainEvent> handleAddTag(Command.AddTagToTask cmd) {
        return List.of(new DomainEvent.TagAddedToTask(
                cmd.taskId().value(),
                Instant.now(),
                cmd.tag()
        ));
    }

    private List<DomainEvent> handleMoveToProject(Command.MoveTaskToProject cmd) {
        return List.of(new DomainEvent.TaskMovedToProject(
                cmd.taskId().value(),
                Instant.now(),
                cmd.projectId()
        ));
    }
}

File: kotlin-app/src/main/kotlin/com/showcase/kotlin/domain/command/CommandHandler.kt
package com.showcase.kotlin.domain.command

import com.showcase.kotlin.domain.bus.EventBus
import com.showcase.kotlin.domain.event.DomainEvent
import com.showcase.kotlin.domain.store.EventStore
import java.time.Instant

class CommandHandler(
    private val eventStore: EventStore,
    private val eventBus: EventBus
) {
    suspend fun handle(command: Command): List<DomainEvent> {
        val events = when (command) {
            is Command.CreateTask -> handleCreate(command)
            is Command.AssignTask -> handleAssign(command)
            is Command.ChangeTaskPriority -> handleChangePriority(command)
            is Command.CompleteTask -> handleComplete(command)
            is Command.AddTagToTask -> handleAddTag(command)
            is Command.MoveTaskToProject -> handleMoveToProject(command)
        }
        eventStore.append(command.aggregateId, events)
        eventBus.publish(events)
        return events
    }

    private fun handleCreate(cmd: Command.CreateTask): List<DomainEvent> {
        val now = Instant.now()
        return listOf(
            DomainEvent.TaskCreated(
                aggregateId = cmd.taskId.value,
                timestamp = now,
                title = cmd.title,
                description = cmd.description,
                priority = cmd.priority,
                dueDate = cmd.dueDate,
                tags = cmd.tags
            )
        )
    }

    private fun handleAssign(cmd: Command.AssignTask): List<DomainEvent> =
        listOf(
            DomainEvent.TaskAssigned(
                aggregateId = cmd.taskId.value,
                timestamp = Instant.now(),
                assigneeId = cmd.assigneeId
            )
        )

    private fun handleChangePriority(cmd: Command.ChangeTaskPriority): List<DomainEvent> =
        listOf(
            DomainEvent.TaskPriorityChanged(
                aggregateId = cmd.taskId.value,
                timestamp = Instant.now(),
                oldPriority = cmd.newPriority, // simplified — real impl would load current priority
                newPriority = cmd.newPriority
            )
        )

    private fun handleComplete(cmd: Command.CompleteTask): List<DomainEvent> {
        val now = Instant.now()
        return listOf(
            DomainEvent.TaskCompleted(
                aggregateId = cmd.taskId.value,
                timestamp = now,
                completedAt = now
            )
        )
    }

    private fun handleAddTag(cmd: Command.AddTagToTask): List<DomainEvent> =
        listOf(
            DomainEvent.TagAddedToTask(
                aggregateId = cmd.taskId.value,
                timestamp = Instant.now(),
                tag = cmd.tag
            )
        )

    private fun handleMoveToProject(cmd: Command.MoveTaskToProject): List<DomainEvent> =
        listOf(
            DomainEvent.TaskMovedToProject(
                aggregateId = cmd.taskId.value,
                timestamp = Instant.now(),
                projectId = cmd.projectId
            )
        )
}
